perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_eduPersonScopedAffiliations.java

```diff
@@ -20,4 +45,73 @@ public String getDestinationAttributeFriendlyName() {
+		if (attribute != null) {
+			Object value = attribute.getValue();
+			if (value != null && value instanceof Set) {
+				Object[] valArray = ((Set) value).toArray();
+				if (valArray instanceof String[]) {
+					for (String val: (String[]) valArray) {
+						valuesWithoutDuplicities.add(val);
+					}
+				}
+			}
+		}
+
+		Attribute attribute1 = null;
+		try {
+			//get value from urn:perun:user:attribute-def:def:eduPersonScopedAffiliationsManuallyAssigned
+			attribute1 = sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, user, SOURCE_ATTRIBUTE_FRIENDLY_NAME_EPSA_MA_URN);
+		} catch (WrongAttributeAssignmentException | AttributeNotExistsException e) {
+			log.error("cannot read " + SOURCE_ATTRIBUTE_FRIENDLY_NAME_EPSA_MA + " of user " + user.getId(), e);
+		}
+		if (attribute1 != null) {
+			Object value = attribute1.getValue();
+			if (value != null && value instanceof Map) {
+				Map<String, String> valueAsMap = (Map<String, String>) value;
+				valuesWithoutDuplicities.addAll(valueAsMap.keySet());
+			}
+		}
+
+		//convert set to list (values in list will be without duplicities)
+		destinationAttribute.setValue(new ArrayList<>(valuesWithoutDuplicities));
+		return destinationAttribute;
+	}
+
+	@Override
+	public List<String> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, String message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<String> resolvingMessages = super.resolveVirtualAttributeValueChange(perunSession, message);
+
+		if (message == null) return resolvingMessages;
+
+		if (allAttributesRemovedForUserExtSource.matcher(message).find() ||
+				removeUserExtSourceAttribute.matcher(message).find() ||
+				setUserExtSourceAttribute.matcher(message).find()) {
+			log.debug("Resolving virtual attribute value change for attribute " + this.SOURCE_ATTRIBUTE_FRIENDLY_NAME_EPSA_MA + " and message: " + message);
+			User user = perunSession.getPerunBl().getModulesUtilsBl().getUserFromMessage(perunSession, message);
+			if (user != null) {
+				Attribute attribute = perunSession.getPerunBl().getAttributesManagerBl().getAttribute(perunSession, user, getDestinationAttributeName());
+				@SuppressWarnings("unchecked") List<String> attributeValue = (ArrayList<String>) attribute.getValue();
+				String messageAttributeSet;
+				if (attributeValue == null || attributeValue.isEmpty()) {
+					AttributeDefinition attributeDefinition = new AttributeDefinition(attribute);
+					messageAttributeSet = attributeDefinition.serializeToString() + " removed for " + user.serializeToString() + ".";
+				} else {
+					messageAttributeSet = attribute.serializeToString() + " set for " + user.serializeToString() + ".";
+				}
+				resolvingMessages.add(messageAttributeSet);
+			}
+			if(!resolvingMessages.isEmpty()) log.debug("These new messages will be generated: " + resolvingMessages);
```

Comment:
I think this method can be fully override, because it looks very similar and there is no really advantage of using the code from the super method.